workflow/sync/sync_manager.go

```diff
@@ -226,61 +271,138 @@ func (cm *SyncManager) ReleaseAll(wf *wfv1.Workflow) bool {
-			} else {
-				semaphoreWaiting.Holders = currentHolder
-				wf.Status.Synchronization.Semaphore.Waiting[index] = semaphoreWaiting
-			}
+	// Update the semaphore which the workflow is waiting for
+	if lockAction == LockActionWaiting {
+		index, semaphoreWaiting := getSemaphoreHolding(wf.Status.Synchronization.Semaphore.Waiting, lockKey)
+		currentHolder := cm.getCurrentLockHolders(lockKey)
+		if index == -1 {
+			wf.Status.Synchronization.Semaphore.Waiting = append(wf.Status.Synchronization.Semaphore.Waiting, wfv1.SemaphoreHolding{Semaphore: lockKey, Holders: currentHolder})
+		} else {
+			semaphoreWaiting.Holders = currentHolder
+			wf.Status.Synchronization.Semaphore.Waiting[index] = semaphoreWaiting
+		}
+		return true
+	}
+	// Update the semaphore which is acquired by the workflow
+	if lockAction == LockActionAcquired {
+		index, semaphoreHolding := getSemaphoreHolding(wf.Status.Synchronization.Semaphore.Holding, lockKey)
+		items := strings.Split(holderKey, "/")
+		holdingName := items[len(items)-1]
+		if index == -1 {
+			wf.Status.Synchronization.Semaphore.Holding = append(wf.Status.Synchronization.Semaphore.Holding, wfv1.SemaphoreHolding{Semaphore: lockKey, Holders: []string{holdingName}})
 			return true
+		} else {
+			if !slice.ContainsString(semaphoreHolding.Holders, holdingName) {
+				semaphoreHolding.Holders = append(semaphoreHolding.Holders, holdingName)
+				wf.Status.Synchronization.Semaphore.Holding[index] = semaphoreHolding
+				return true
+			}
+		}
+		return false
+	}
+	// Clear the semaphore which is released by the workflow
+	if lockAction == LockActionReleased {
+		items := strings.Split(holderKey, "/")
+		holdingName := items[len(items)-1]
+		index, semaphoreHolding := getSemaphoreHolding(wf.Status.Synchronization.Semaphore.Holding, lockKey)
+		if index != -1 {
+			semaphoreHolding.Holders = slice.RemoveString(semaphoreHolding.Holders, holdingName)
+			wf.Status.Synchronization.Semaphore.Holding[index] = semaphoreHolding
 		}
+		return true
+	}
+	return false
+}
 
-		if lockAction == LockActionAcquired {
-			index, semaphoreHolding := getSemaphoreHolding(wf.Status.Synchronization.Semaphore.Holding, lockKey)
-			items := strings.Split(holderKey, "/")
-			holdingName := items[len(items)-1]
-			if index == -1 {
-				wf.Status.Synchronization.Semaphore.Holding = append(wf.Status.Synchronization.Semaphore.Holding, wfv1.SemaphoreHolding{Semaphore: lockKey, Holders: []string{holdingName}})
-				return true
-			} else {
-				if !slice.ContainsString(semaphoreHolding.Holders, holdingName) {
-					semaphoreHolding.Holders = append(semaphoreHolding.Holders, holdingName)
-					wf.Status.Synchronization.Semaphore.Holding[index] = semaphoreHolding
-					return true
-				}
+// updateMutexStatus updates the mutex holding and waiting details
+func (cm *SyncManager) updateMutexStatus(holderKey, lockKey string, lockAction LockAction, wf *wfv1.Workflow) bool {
```

Comment:
I was thought this but both are totally different data structures and functionality.  Semaphore needs to manipulate with an array of elements (check, add, delete) but in mutex has only one item and just replace it.  It will more complex function (with multiple ifs) if we merge both in one function